ingestion-edge/bin/test

```diff
@@ -0,0 +1,98 @@
+  --lint-only           only run lint tests (-k 'not test_'), cover tests/load
+                        by default (-o 'testpaths='), add lint options (--lint)
+  --no-clean-relocates  do not remove .pyc files that appear to be relocated
+  --no-pytest           do not call pytest
+  --no-venv             do not use a venv to run tests"
+
+function argparse() {
+    while ! (exit $#); do
+        # handle arg
+        case "$1" in
+            -h | --help )
+                # print USAGE and pytest help
+                HELP=true ARGS=( "${ARGS[@]}" --help ) ;;
+            --build )
+                # create or update virtualenv
+                BUILD=true ;;
+            --lint )
+                # add lint options
+                ARGS=( "${ARGS[@]}" "${LINT[@]}" ) ;;
+            --lint-only )
+                # add lint options
+                argparse --lint
+                # only run plugin (lint) tests
+                ARGS=( "${ARGS[@]}" -k "not test_" )
+                # cover tests/load by default
+                ARGS=( "${ARGS[@]}" -o "testpaths=" ) ;;
+            --no-clean-relocates )
+                # do not remove .pyc files that appear to be relocated
+                CLEAN_RELOCATES=false ;;
+            --no-pytest )
+                # do not run pytest
+                PYTEST=false ;;
+            --no-venv )
+                # do not use venv
+                VENV=false ;;
+            -- )
+                # pass through remaining args and break
+                ARGS=( "${ARGS[@]}" "$@" )
+                break
+                ;;
+            * )
+                # pass through unmatched arg
+                ARGS=( "${ARGS[@]}" "$1" )
+        esac
+        # remove arg from $@
+        shift
+    done
+}
+
+# parse "$@" into "${ARGS[@]}" with default arg --lint
+argparse "${@:---lint}"
+
+# use venv unless disabled
+if ${VENV:-true}; then
+    VENV_DIR="$PWD/venv/$(uname)"
+    python3.7 -m venv "$VENV_DIR"
+    PATH="$VENV_DIR/bin:$PATH"
+fi
+
+# install or update dependencies if requested
+if ${BUILD:-false}; then
+    python3.7 -m pip install --upgrade --no-cache-dir -r requirements.txt
+    # remove after merge and release of https://github.com/mozilla-services/python-dockerflow/pull/27
+    python3.7 -m pip install --no-deps --no-index git+git://github.com/relud/python-dockerflow@master#egg=dockerflow[sanic]
+fi
+
+# remove .pyc files that appear to be relocated unless disabled
+if ${CLEAN_RELOCATES:-true}; then
+    # list .pyc files that do not match "$PWD/"
+    RELOCATES="$(grep -rL "$PWD/" --include '*.pyc' --exclude-dir venv . || true)"
+    # remove RELOCATES if found
+    test -z "$RELOCATES" || echo "$RELOCATES" | xargs rm
+fi
+
+# print USAGE on --help
+! ${HELP:-false} || echo "$USAGE"
+
+# exit early on --no-pytest
+${PYTEST:-true} || exit 0
+
+if ${HELP:-false}; then
+    python3.7 -m pytest "${ARGS[@]}" | sed 1d
+else
+    exec python3.7 -m pytest "${ARGS[@]}"
+fi
```

Comment:
Did you use some sort of framework for generating `bin/test` script? It's some pretty advanced bash option parsing. I find it quite readable, but it _is_ rather complex so I have some worry about maintenance burden. If someone fresh to the repo had to add a new option, for example, it might take a while to get up to speed on how to do so.

I was imagining a thinner wrapper that would just call a single line like `pytest --black --bunch-of-other-shared-args $@` but it looks like that may not be flexible enough to meet the needs of the various invocations here.

If this is building on some existing example, I'd feel better if we linked to that prior art and were able to reference that to incorporate future updates, etc.

If this is entirely new code, I rate it a _wow_ 😮 for high cleanliness and level of documentation for a bash script.

Reply:
it's entirely new, but it's not my first time doing option parsing in bash 😊

Reply:
i'll sit on this and think of ways to reduce complexity

Reply:
@jklukas okay, what do you think of 24ea67e ? I need to document `VENV` and `CLEAN_RELOCATES` somewhere, but i'm not sure where best to do it.

Reply:
I do think the refactor here is more maintainable. I would say VENV and `CLEAN_RELOCATES` can be documented in the README where you give example invocations of the `bin/` scripts.